Fix false positives in action selector

habbes opened the pull request:
<!-- markdownlint-disable MD002 MD041 -->

### Issues

*This pull request fixes issue #1973*

### Description

*Briefly describe the changes of this pull request.*

I updated the `ODataActionSelector` in both AspNetCore (2.x and 3.0) and AspNet classic and made them more strict. It rejects candidate actions which have parameters which cannot be matched to the route parameters, body, query options, optional parameters and `ODataPath`.

### Checklist (Uncheck if it is not completed)

- [x] *Test cases added*
- [x] *Build and test with one-click build and test script passed*

(PS: the build command seems to fail due to a recent library update that added support for .net 3.0, not due to my PR).

### Additional work necessary

If this PR is accepted, it would be a good idea to update the docs mentioning that a 404 would be returned if the parameters of the action don't match those of the request, even if the name matches.

odero commented on test/UnitTest/Microsoft.AspNet.OData.Test.Shared/Routing/ODataRoutingTest.cs:
```diff
@@ -893,4 +911,40 @@ public string GetName(int keyID)
+    {
+        public string Get(int key)
+        {
+            return String.Format(CultureInfo.InvariantCulture, "Get({0})", key);
```
```suggestion
            return $"Get({key})");
```

habbes commented on src/Microsoft.AspNetCore.OData/Routing/ODataActionSelector.cs:
```diff
@@ -80,10 +81,23 @@ public ActionDescriptor SelectBestCandidate(RouteContext context, IReadOnlyList<
+                var _odataRoute = router as ODataRoute;
+                return _odataRoute == null ?
+                    false : true;
+            }) as ODataRoute;
```
`routeData.Routers` happens to have objects of different types that implement `IRouter`. Only one of them was an `ODataRoute` type, and that's the one that has the `RoutePrefix` property.

odero replied:
Untested but maybe something along these lines then?
```c#
ODataRoute odataRoute = routeData.Routers.Where(router => router is ODataRoute).FirstOrDefault();
```

xuzhg commented on src/Microsoft.AspNetCore.OData/Routing/ODataActionSelector.cs:
```diff
@@ -80,10 +81,23 @@ public ActionDescriptor SelectBestCandidate(RouteContext context, IReadOnlyList<
             ODataPath odataPath = context.HttpContext.ODataFeature().Path;
+
+            var routers = routeData.Routers;
+            var odataRoute = routeData.Routers.FirstOrDefault((router) =>
```
>routeData.Routers [](start = 29, length = 17)

it's basic scenario to register multiple ODataRoute in the application. It's supported in Web API OData.

habbes replied:
Thanks for the clarification.

xuzhg commented on src/Microsoft.AspNetCore.OData/Routing/ODataActionSelector.cs:
```diff
@@ -80,10 +81,23 @@ public ActionDescriptor SelectBestCandidate(RouteContext context, IReadOnlyList<
+                var _odataRoute = router as ODataRoute;
+                return _odataRoute == null ?
+                    false : true;
+            }) as ODataRoute;
```
You can accomplish it:
1) routeData.Routers.OfType<ODataRoute>().FirstOrDefault()
2) routeData.Routers.FirstOrDefault(r => r is ODataRoute) as ODataRoute;

xuzhg commented on test/UnitTest/Microsoft.AspNetCore.OData.Test/Abstraction/ControllerDescriptorFactory.cs:
```diff
@@ -8,6 +8,8 @@
 using Microsoft.AspNetCore.Routing;
 using Microsoft.Extensions.DependencyInjection;
+using System.Linq;
+using Microsoft.AspNetCore.Mvc.Abstractions;
```
sorting the using

xuzhg commented on test/UnitTest/Microsoft.AspNetCore.OData.Test/Abstraction/ODataActionSelectorTestHelper.cs:
```diff
@@ -0,0 +1,78 @@
+
+namespace Microsoft.AspNet.OData.Test.Abstraction
+{
+    class ODataActionSelectorTestHelper
```
>class [](start = 4, length = 5)

explicitly decorate it as internal

xuzhg commented on src/Microsoft.AspNetCore.OData/Routing/ODataActionSelector.cs:
```diff
@@ -80,10 +81,23 @@ public ActionDescriptor SelectBestCandidate(RouteContext context, IReadOnlyList<
                 // Get the available parameter names from the route data. Ignore case of key names.
-                IList<string> availableKeys = routeData.Values.Keys.Select(k => k.ToLowerInvariant()).AsList();
+                IList<string> availableKeys = routeData.Values.Keys.Select(k => k.ToLowerInvariant())
+                    .Where((key) => routePrefix != "{" + key + "}")
```
>(key) => routePrefix != "{" + key + "}") [](start = 27, length = 40)

what's the meaning to compare the {key} with the RoutePrefix ?

habbes replied:
I have found that when the Web API OData is configured with a parameterized prefix, that prefix is included in the available keys. So in that statement I was removing the prefix from the available keys in order to easily detect when a candidate action has no parameters to match the actual keys in the path. But this does feel like a hack and there's probably a better or clearer way to achieve the goal. Or maybe I just misunderstand the use case of a parameterized route prefix. Why is the route prefix added to the route data, is it meant to be bound to action parameters?

habbes commented:
Hi @xuzhg @odero, here are some updates on the task:

I've updated my branch with the latest updates from master, including support for .net core 3.0. My ODataActionsSelector (AspNet Core) tests were failing to build because the ODataActionSelector constructor I was using is not supported in 3.0. I'll work on fixing these tests. Alternatively I could just get rid of these tests since I also added other more general tests in ODataRoutingTest (shared by both core and classic) that test against the expected routing behaviour.

When running the tests under `netcoreapp3` environment, some tests fail due to how I was testing whether the request contains a body. I was using `context.HttpContext.Request.Body.Length`. This worked in `netcoreapp2`. But under `netcoreapp3` tests, the request body is not seekable and therefore getting the length throws an exception.

To provide some context, the reason I was checking whether the request has a body, is to see whether an action parameter that is not bound to a key in the path is bound to the body. This helped match actions like `Post(Person body)` to routes like `POST /Persons` where the parameter does not match any segment in the route, but will receive its value from the request body. On the other hand,
if the action has parameters that are neither in the route segments or in the body, then a 404 should be returned. For example, if the only controller action is `Get(int key)` and the route is `GET /Persons` then a 404 is returned.

Using the request body length allowed me to achieve that behaviour. But probably that's a hack. Kindly advise if you have a better approach.

I'm still working on an implementaton for AspNet classic.

habbes commented:
I've made slight adjustments to my AspNetCore `ODataActionSelector` changes and now they build and pass tests in both net core 3.0 and 2.0. I used conditional compiler directives in some tests for this to work.

I've also ported the changes I've made over to the AspNet classic project. However, there's one pre-existing unit test that fails in AspNet classic as a result of my changes, I have not fixed this yet because I wanted to get feedback on my approach first, because this test seems to break the assumptions I had about how routing is supposed to behave.

What I did was to take the existing `ODataActionSelector.SelectBestCandidate` and `TryMatch` methods and made them stricter in the following ways:
All the parameters of the selected action have to either:
- match a key in the route data values
- be an optional parameter
- match query options
- be of type `ODataPath`
- be bound to the request body

If the method has extra parameters that do not match any of the above conditions, then that method is rejected (i.e. `Get(key)` will not get selected for`/odata/Entities`). If the method has no parameters but the route data has values (e.g. `key` or `relatedkey`), then the method is rejected (this means that `Get()` won't get selected when the route is `/odata/Entities(key)`). With these adjustments, the
selector will return `null` (i.e. 404 error) when no action matches the conditions.

The `ODataActionSelector` assumes that a parameter will be bound to the body if the parameter does not match any of the other conditions and if the request has a request body (as determined by the `Content-Length` header). Since binding happens later, this was the best estimate I could come up with.

In cases where the odata route has a parameterized prefix, this route prefix is added to the route data values. Is this something I should account for? Is there a case where this would be bound to parameters? Currently all tests pass when I don't include it in the conditions I listed above.

These additional checks and conditions ensured that the I eliminated the behavior described in the issue, and also kept existing tests* passing.

The test that fails is enabled in AspNet classic but not in Core, namely `CanModelBindNonStringData()`  in `Microsoft.AspNet.OData.Test.Shared\Routing\ODataValueProviderFactoryTest.cs`.
It calls the route `http://localhost/Test`, manually adds some data to the request's `RoutingConventionStore` and expects it to call `Get(TestClass test)` with the data bound to the `test` parameter.

This does not work with my update because `Get(test)` has a parameter that's neither in the route nor in the body. My update assumes that what the test is trying to do is invalid. I could probably make the test pass simply by adding the parameter name to its route data. But I first would like to confirm that my assumption is correct and that in the real application This argument would either come from the route or the body.

cc @xuzhg @mikepizzo @odero

habbes commented:
I've fixed the failing test in AspNet classic by updating the `ODataActionSelector` to look for the parameter in the `RoutingConventionsStore` as well.

xuzhg commented on test/UnitTest/Microsoft.AspNetCore.OData.Test/Routing/ODataActionSelectorTest.cs:
```diff
@@ -0,0 +1,213 @@
+using Microsoft.AspNet.OData.Test.Common;
+using System.Collections.Generic;
+using Xunit;
+using System;
```
sorting
System.* at head

xuzhg commented on test/UnitTest/Microsoft.AspNetCore.OData.Test/Routing/ODataActionSelectorTest.cs:
```diff
@@ -0,0 +1,213 @@
+        {
+            get
+            {
+                return new TheoryDataSet<Dictionary<string, object>, string, System.Type, string, System.Type[]>
```
>System. [](start = 98, length = 7)

can get rid of System.

xuzhg commented:
Now, you can run the build.cmd netfx|netcore on command line. please let me know any problems. 

---
In reply to: [568395516](https://github.com/OData/WebApi/pull/1988#issuecomment-568395516) [](ancestors = 568395516)

habbes commented:
@xuzhg I've updated my code to work with .Net Core 3.1 following the recent updates to the WebApi library. My branch now builds correctly when I run `build.cmd quick`. However, it fails on Azure pipelines because: `The current .NET SDK does not support targeting .NET Core 3.1`.

habbes commented:
This PR fails to build on CI because it's using an old build agent. To resolve this, I have open a separate PR checkout from this branch.

Let's the conversation continue over on this PR: https://github.com/OData/WebApi/pull/2007

odero commented:
@habbes does that mean this PR can be closed?

habbes commented:
@odero yeah, I guess it makes sense to close the PR.
